### DISCUSSION TITLE
How to Add FancyIndex

### DISCUSSION THREAD

**Discussion by ajschmidt8** [nginxinc/nginx-s3-gateway#33] [Q&A] [answered]
Can someone provide a summary of how I can incorporate the `FancyIndex` module into these images? For reference, I'm using regular `nginx` and not `nginx-plus`. Thanks!

**Comment by ajschmidt8** [comment]
Would it be similar to this, where you build `fancyindex` as a dynamic module and then use it at runtime?

- https://gist.github.com/Dimtree/d07a537bcdfa6fe672b99ad08d93ceee

**Comment by dekobon** [comment]
Hi @ajschmidt8 , are you trying to get a fancy index style of listing for S3 bucket listings?

**Reply by ajschmidt8** [reply]
Yes, that's correct. I was wondering if I could just compile the `fancyindex` dynamically and then extend the `nginx-s3-gateway` images to include the `ngx_http_fancyindex_module.so` in `/etc/nginx/modules`. Does that seem like it would work?

**Reply by dekobon** [reply]
That won't work because fancy index is written so that it works with files that are coming off of the file system. If you want to modify how files are displayed from S3 listings, all you need to do is modify the [XSL template](https://www.w3schools.com/xml/xsl_intro.asp) that renders them into HTML. The file is here: https://github.com/nginxinc/nginx-s3-gateway/blob/master/common/etc/nginx/include/listing.xsl

**Reply by dekobon** [reply]
If you come up with a nice design, I'd be happy to have it part of this repository.

**Reply by ajschmidt8** [reply]
@dekobon, I see. My main motive was to get human-readable file sizes to display on the directory pages instead of displaying file sizes in bytes. I see that the `autoindex` module [here](https://nginx.org/en/docs/http/ngx_http_autoindex_module.html) also supports the `autoindex_exact_size` option like `fancyindex`, but I suspect that it suffers from the same issue you mentioned above about the filesystem. Is that correct? Is there an alternative way that I could get human-readable file sizes displayed for directory listings?

**Reply by dekobon** [reply]
Perhaps you could use [filesizejs](https://filesizejs.com) to render the raw byte sizes into a human readable form using Javascript? The XSLT stylesheet is just something that converts an input XML file to HTML. You can edit it and embed in Javascript and CSS to your heart's content.

**Reply by ajschmidt8** [reply]
Okay. That seems like a decent plan. Thanks!

**Comment by nathanweeks** [comment]
Would it be feasible to use [aws-s3-js-explorer](https://github.com/awslabs/aws-js-s3-explorer)? It might require the [v2-alpha branch to use a custom domain](https://github.com/awslabs/aws-js-s3-explorer/issues/29), and the source might need light modification (at least, it didn't seem to work out-of-the-box for my custom domain).

**Reply by dekobon** [reply]
Interesting. That might work. Someone should try and see how it works in combination with the nginx gateway.

**Reply by nathanweeks** [reply]
It doesn't look like aws-s3-js-explorer is compatible, as it attempts to use the AWS S3 REST API, and the nginx S3 gateway appears to remove the query string from the request? https://github.com/nginxinc/nginx-s3-gateway/blob/35681732dcb4111559469edc54873af31bbc52d1/common/etc/nginx/templates/default.conf.template#L7-L12 ?

If that's correct, would there be a straightforward way of changing the behavior of nginx-s3-gateway to allow the AWS S3 REST API to be used, or is its current behavior fundamental to how it operates?

**Reply by dekobon** [reply]
I believe that is correct - it wouldn't be compatible, unless you could point it directly at a S3 api without assuming the same host as the requested page. The gateway is really nothing more than a collection of NGINX configuration, so you are free to modify it for what works for you.

**Comment by dekobon** [comment, marked as answer]
@ajschmidt8 Please keep us up to date on what your final solution is.

**Reply by ajschmidt8** [reply]
I've opted to use filesizejs to do the conversion on the client side. Appreciate the recommendation!
